src/ram/transform/MakeIndex.cpp

```diff
@@ -38,131 +38,225 @@
-                    return {clone(lhs), std::make_unique<RamUndefValue>()};
-                }
+        }
+        // <expr> >= Tuple[level, element]
+        if (const auto* rhs = dynamic_cast<const RamTupleElement*>(&binRelOp->getRHS())) {
+            const RamExpression* lhs = &binRelOp->getLHS();
+            if (rhs->getTupleId() == identifier && rla->getLevel(lhs) < identifier) {
+                element = rhs->getElement();
+                return {std::make_unique<RamUndefValue>(), clone(lhs)};
             }
         }
+    }
+    if (isLessThanSigned(binRelOp->getOperator())) {
+        // Tuple[level, element] < <expr>
+        // is equivalent to
+        // Tuple[level, element] <= <expr> - 1
+        if (const auto* lhs = dynamic_cast<const RamTupleElement*>(&binRelOp->getLHS())) {
+            const RamExpression* rhs = &binRelOp->getRHS();
+            if (lhs->getTupleId() == identifier && rla->getLevel(rhs) < identifier) {
+                element = lhs->getElement();
+                std::vector<std::unique_ptr<RamExpression>> expressions;
+                expressions.push_back(std::unique_ptr<RamExpression>(rhs->clone()));
+                expressions.push_back(std::make_unique<RamSignedConstant>(RamDomain(1)));
 
-        if (isGreaterEqualConstraint(binRelOp->getOperator())) {
-            // Tuple[level, element] >= <expr>
-            if (const auto* lhs = dynamic_cast<const RamTupleElement*>(&binRelOp->getLHS())) {
-                const RamExpression* rhs = &binRelOp->getRHS();
-                if (lhs->getTupleId() == identifier && rla->getLevel(rhs) < identifier) {
-                    element = lhs->getElement();
-                    return {clone(rhs), std::make_unique<RamUndefValue>()};
-                }
+                return {std::make_unique<RamUndefValue>(),
+                        std::make_unique<RamIntrinsicOperator>(FunctorOp::SUB, std::move(expressions))};
             }
-            // <expr> >= Tuple[level, element]
-            if (const auto* rhs = dynamic_cast<const RamTupleElement*>(&binRelOp->getRHS())) {
-                const RamExpression* lhs = &binRelOp->getLHS();
-                if (rhs->getTupleId() == identifier && rla->getLevel(lhs) < identifier) {
-                    element = rhs->getElement();
-                    return {std::make_unique<RamUndefValue>(), clone(lhs)};
-                }
+        }
+        // <expr> < Tuple[level, element]
+        // is equivalent to
+        // <expr> + 1 <= Tuple[level, element]
+        if (const auto* rhs = dynamic_cast<const RamTupleElement*>(&binRelOp->getRHS())) {
+            const RamExpression* lhs = &binRelOp->getLHS();
+            if (rhs->getTupleId() == identifier && rla->getLevel(lhs) < identifier) {
+                element = rhs->getElement();
+                std::vector<std::unique_ptr<RamExpression>> expressions;
+                expressions.push_back(std::unique_ptr<RamExpression>(lhs->clone()));
```

Comment:
```suggestion
                expressions.push_back(souffle::clone(lhs));
```
or probably just `clone()`, since the transformers don't have their own clone methods so it's safe here.